cmd/kindnetd/main.go

```diff
@@ -0,0 +1,238 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Note: the example only works with the code within the same release/branch.
+package main
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"text/template"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netlink/nl"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// kindnetd is a simple networking daemon to complete kind's CNI implementation
+// kindnetd will ensure routes to the other node's PodCIDR via their InternalIP
+// kindnetd will also write a templated cni config supplied with PodCIDR
+//
+// input envs:
+// - HOST_IP: hould be populated by downward API
+// - POD_IP: should be populated by downward API
+// - CNI_CONFIG_TEMPLATE: the cni .conflist template, run with {{ .PodCIDR }}
+
+// TODO: improve logging & error handling
+
+func main() {
+	// create a client
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		panic(err.Error())
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	// obtain the host and pod ip addresses
+	// if both ips are different we are not using the host network
+	hostIP, podIP := os.Getenv("HOST_IP"), os.Getenv("POD_IP")
+	fmt.Printf("hostIP = %s\npodIP = %s\n", hostIP, podIP)
+	if hostIP != podIP {
+		panic(fmt.Sprintf(
+			"hostIP(= %q) != podIP(= %q) but must be running with host network: ",
+			hostIP, podIP,
+		))
+	}
+
+	// used to track if the cni config inputs changed and write the config
+	cniConfigWriter := &CNIConfigWriter{
+		path:     CNIConfigPath,
+		template: os.Getenv("CNI_CONFIG_TEMPLATE"),
+	}
+	// setup nodes reconcile function, closes over arguments
+	reconcileNodes := makeNodesReconciler(cniConfigWriter, hostIP)
+
+	// main control loop
+	for {
+		// Gets the Nodes information from the API
+		nodes, err := clientset.CoreV1().Nodes().List(metav1.ListOptions{})
+		if err != nil {
+			panic(err.Error())
+		}
+
+		// reconcile the nodes
+		if err := reconcileNodes(nodes); err != nil {
+			panic(err.Error())
+		}
+
+		// rate limit
+		time.Sleep(10 * time.Second)
+	}
+}
+
+// nodeNodesReconciler returns a reconciliation func for nodes
+func makeNodesReconciler(cniConfigWriter *CNIConfigWriter, hostIP string) func(*corev1.NodeList) error {
+	// reconciles a node
+	reconcileNode := func(node corev1.Node) error {
+		// first get this node's IP
+		nodeIP := internalIP(node)
+		fmt.Printf("Handling node with IP: %s\n", nodeIP)
+
+		// This is our node. We don't need to add routes, but we might need to
+		// update the cni config.
+		if nodeIP == hostIP {
+			fmt.Printf("handling current node\n")
+			// compute the current cni config inputs
+			if err := cniConfigWriter.Write(
```

Comment:
are we continue writing the file or only if something has changed_

Reply:
only if something changed, the writer handles this 😅